core/test/base/range_accessors.cpp

```diff
@@ -121,4 +139,569 @@ TEST_F(RowMajorAccessor, CanAssignSubranges)
+        if (ignore != t(1, 1, 0)) { EXPECT_EQ(a(1, 1, 0), 2.08);    }
+        if (ignore != t(1, 1, 1)) { EXPECT_EQ(a(1, 1, 1), 3.09);    }
+        if (ignore != t(1, 2, 0)) { EXPECT_EQ(a(1, 2, 0), -1.1);    }
+        if (ignore != t(1, 2, 1)) { EXPECT_EQ(a(1, 2, 1), -9.11);   }
+        if (ignore != t(2, 0, 0)) { EXPECT_EQ(a(2, 0, 0), -2.12);   }
+        if (ignore != t(2, 0, 1)) { EXPECT_EQ(a(2, 0, 1), 2.13);    }
+        if (ignore != t(2, 1, 0)) { EXPECT_EQ(a(2, 1, 0), 0.14);    }
+        if (ignore != t(2, 1, 1)) { EXPECT_EQ(a(2, 1, 1), 15.15);   }
+        if (ignore != t(2, 2, 0)) { EXPECT_EQ(a(2, 2, 0), -9.16);   }
+        if (ignore != t(2, 2, 1)) { EXPECT_EQ(a(2, 2, 1), 8.17);    }
+        if (ignore != t(3, 0, 0)) { EXPECT_EQ(a(3, 0, 0), 7.18);    }
+        if (ignore != t(3, 0, 1)) { EXPECT_EQ(a(3, 0, 1), -6.19);   }
+        if (ignore != t(3, 1, 0)) { EXPECT_EQ(a(3, 1, 0), 5.2);     }
+        if (ignore != t(3, 1, 1)) { EXPECT_EQ(a(3, 1, 1), -4.21);   }
+        if (ignore != t(3, 2, 0)) { EXPECT_EQ(a(3, 2, 0), 3.22);    }
+        if (ignore != t(3, 2, 1)) { EXPECT_EQ(a(3, 2, 1), -2.23);   }
+        // clang-format on
+    }
+};
+
+
+TEST_F(ReducedStorage3d, CorrectLengths)
+{
+    EXPECT_EQ(r.length(0), size[0]);
+    EXPECT_EQ(r.length(1), size[1]);
+    EXPECT_EQ(r.length(2), size[2]);
+    EXPECT_EQ(r.length(3), 1);
+    EXPECT_EQ(r->get_size(), size);
+}
+
+
+TEST_F(ReducedStorage3d, CorrectStride)
+{
+    EXPECT_EQ(r->get_stride()[0], size[1] * size[2]);
+    EXPECT_EQ(r->get_stride().at(0), size[1] * size[2]);
+    EXPECT_EQ(r->get_stride()[1], size[2]);
+    EXPECT_EQ(r->get_stride().at(1), size[2]);
+}
+
+
+TEST_F(ReducedStorage3d, CorrectStorage)
+{
+    EXPECT_EQ(r->get_storage(), data);
+    EXPECT_EQ(r->get_const_storage(), data);
+}
+
+
+TEST_F(ReducedStorage3d, CanReadData)
+{
+    check_accessor_correctness(r);
+    check_accessor_correctness(cr);
+}
+
+
+TEST_F(ReducedStorage3d, CopyFrom)
+{
+    st_type data2[data_elements];
+    reduced_storage cpy(data2, size);
+
+    // Do not use this in regular code since the implementation is slow
+    cpy = r;
+
+    check_accessor_correctness(cpy);
+}
+
+
+TEST_F(ReducedStorage3d, CanImplicitlyConvertToConst)
+{
+    const_reduced_storage const_rs = r->to_const();
+    const_reduced_storage const_rs2 = cr;
+
+    check_accessor_correctness(const_rs);
+    check_accessor_correctness(const_rs2);
+}
+
+
+TEST_F(ReducedStorage3d, ToConstWorking)
+{
+    auto cr2 = r->to_const();
+
+    static_assert(std::is_same<decltype(cr2), const_reduced_storage>::value,
+                  "Types must be equal!");
+    check_accessor_correctness(cr2);
+}
+
+
+TEST_F(ReducedStorage3d, CanWriteData)
+{
+    r(0, 1, 0) = 100.2;
+
+    check_accessor_correctness(r, t(0, 1, 0));
+    EXPECT_EQ(r(0, 1, 0), 100.2);
+}
+
+
+TEST_F(ReducedStorage3d, Assignment)
+{
+    r(0, 0, 1) = 10.2;
+
+    check_accessor_correctness(r, t(0, 0, 1));
+    EXPECT_NEAR(r(0, 0, 1), 10.2, delta);
```

Comment:
Yes, but is not necessarily for `double` -> `float`. I was thinking about making this test for multiple precisions, and I going through `static_cast` twice is not guaranteed to result in the same value (I might have to change the `check_accessor_correctness` in that regard as well, but it is centralized, so changing that is easy).

Reply:
Maybe then you can set `delta` to 0.0 for the functions where it is supposed to be exact ?

Reply:
The big question for that is if we want to have this test for multiple precisions, or not. If we don't, then testing for equality is fine. If we do want to test multiple types, we can set `delta` to `0.0` in case storage and arithmetic types are identical.

Reply:
I think it makes sense to have tests for multiple precisions and set `delta` to zero when the storage and arithmetic types are the same.